cpp/src/neighbors/ivf_flat/ivf_flat_interleaved_scan.cuh

```diff
@@ -202,335 +217,356 @@ struct loadAndComputeDist {
+
+// This handles uint8_t 8, 16 Veclens (also handles int8_t via byte_arithmetic_ptr with int32
+// accumulator)
 template <int kUnroll, typename Lambda, int uint8_veclen, bool ComputeNorm>
-struct loadAndComputeDist<kUnroll, Lambda, uint8_veclen, uint8_t, uint32_t, ComputeNorm> {
+struct loadAndComputeDist<kUnroll, Lambda, uint8_veclen, uint8_t, int32_t, ComputeNorm> {
   Lambda compute_dist;
-  uint32_t& dist;
-  uint32_t& norm_query;
-  uint32_t& norm_data;
+  int32_t& dist;
+  int32_t& norm_query;
+  int32_t& norm_data;
 
   __device__ __forceinline__
-  loadAndComputeDist(uint32_t& dist, Lambda op, uint32_t& norm_query, uint32_t& norm_data)
+  loadAndComputeDist(int32_t& dist, Lambda op, int32_t& norm_query, int32_t& norm_data)
     : dist(dist), compute_dist(op), norm_query(norm_query), norm_data(norm_data)
   {
   }
 
-  __device__ __forceinline__ void runLoadShmemCompute(const uint8_t* const& data,
+  __device__ __forceinline__ void runLoadShmemCompute(const cuvs::detail::byte_arithmetic_ptr& data,
                                                       const uint8_t* query_shared,
                                                       int loadIndex,
                                                       int shmemIndex)
   {
-    constexpr int veclen_int = uint8_veclen / 4;  // converting uint8_t veclens to int
-    loadIndex                = loadIndex * veclen_int;
+    constexpr int veclen_int     = uint8_veclen / 4;  // converting uint8_t veclens to int
+    loadIndex                    = loadIndex * veclen_int;
+    const bool is_signed         = data.is_signed;
+    constexpr int32_t offset_reg = 0x80808080;  // 128 in each byte position
 #pragma unroll
     for (int j = 0; j < kUnroll; ++j) {
       uint32_t encV[veclen_int];
-      raft::ldg(
-        encV,
-        reinterpret_cast<unsigned const*>(data) + loadIndex + j * kIndexGroupSize * veclen_int);
+      const int byte_offset = (loadIndex + j * kIndexGroupSize * veclen_int) * 4;
+#pragma unroll
+      for (int i = 0; i < veclen_int; i++) {
+        encV[i] = pack_bytes_uint32(data[byte_offset + i * 4 + 0],
+                                    data[byte_offset + i * 4 + 1],
+                                    data[byte_offset + i * 4 + 2],
+                                    data[byte_offset + i * 4 + 3]);
+      }
       uint32_t queryRegs[veclen_int];
       raft::lds(queryRegs,
                 reinterpret_cast<unsigned const*>(query_shared + shmemIndex) + j * veclen_int);
 #pragma unroll
       for (int k = 0; k < veclen_int; k++) {
-        compute_dist(dist, queryRegs[k], encV[k]);
-        if constexpr (ComputeNorm) {
-          norm_query = raft::dp4a(queryRegs[k], queryRegs[k], norm_query);
-          norm_data  = raft::dp4a(encV[k], encV[k], norm_data);
+        if constexpr (is_inner_prod_dist<Lambda>::value) {
+          int32_t q = is_signed ? (static_cast<int32_t>(queryRegs[k]) - offset_reg)
```

Comment:
Is a `static_cast` correct here? That would change the underlying bit string, isn't it?